src/transformers/models/qwen2_vl/modeling_qwen2_vl.py

```diff
@@ -484,6 +488,156 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+
+    position_ids = _kwargs["position_ids"]
+
+    attn_output = _flash_attention_forward(
+        query_states,
+        key_states,
+        value_states,
+        mask,
+        query_states.shape[1],
+        position_ids=position_ids,
+        dropout=dropout_rate,
+        sliding_window=config.sliding_window,
+        is_causal=config.is_causal,
+        use_top_left_mask=config._flash_attn_uses_top_left_mask,
+    )
+
+    attn_output = attn_output.to(input_dtype)
+    return attn_output, None
+
+
+def sdpa_attention_forward(
+    config: Qwen2VLConfig,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    mask: Optional[torch.Tensor],
+    **_kwargs,
+) -> Tuple[torch.Tensor, None]:
+    key_states = repeat_kv(key, config.num_key_value_groups)
+    value_states = repeat_kv(value, config.num_key_value_groups)
+
+    causal_mask = mask
+    if mask is not None:  # no matter the length, we just slice it
+        causal_mask = mask[:, :, :, : key_states.shape[-2]]
+
+    # SDPA with memory-efficient backend is currently (torch==2.1.2) bugged with non-contiguous inputs with custom attn_mask,
+    # Reference: https://github.com/pytorch/pytorch/issues/112577.
+    if query.device.type == "cuda" and mask is not None:
+        query = query.contiguous()
+        key_states = key_states.contiguous()
+        value_states = value_states.contiguous()
+
+    # We dispatch to SDPA's Flash Attention or Efficient kernels via this `is_causal` if statement instead of an inline conditional assignment
+    # in SDPA to support both torch.compile's dynamic shapes and full graph options. An inline conditional prevents dynamic shapes from compiling.
+    is_causal = True if causal_mask is None and query.shape[1] > 1 else False
+
+    attn_output = torch.nn.functional.scaled_dot_product_attention(
+        query,
+        key_states,
+        value_states,
+        attn_mask=causal_mask,
+        dropout_p=config.attention_dropout if config.training else 0.0,
+        is_causal=is_causal,
+    )
+    attn_output = attn_output.transpose(1, 2).contiguous()
+    return attn_output, None
+
+
+def flex_attention_forward(
+    config: Qwen2VLConfig,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    mask: Optional[torch.Tensor],
+    **_kwargs,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    def causal_mod(score, b, h, q_idx, kv_idx):
+        if mask is not None:
+            score += mask[b][0][q_idx][kv_idx]
+        return score
+
+    attn_output, attn_weights = flex_attention(
+        query,
+        key,
+        value,
+        score_mod=causal_mod,
+        enable_gqa=True,
+        return_lse=True,
+    )
+    attn_weights = attn_weights.to(value.dtype)
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, attn_weights
+
+
+QWEN2_VL_ATTENTION_FUNCTION = {
+    "flash_attention_2": flash_attention_forward,
+    "flex_attention": flex_attention_forward,
```

Comment:
This doesn't really work yet. I believe we need to refactor the vision attentions and add flex attention too.